ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -163,43 +174,81 @@ forkSyncStateOnTipPointChange registry menv =
 
 implAddTxs :: forall m blk. (MonadSTM m, ApplyTx blk)
            => MempoolEnv m blk
+           -> [(GenTx blk, Maybe (ApplyTxErr blk))]
            -> [GenTx blk]
            -> m [(GenTx blk, Maybe (ApplyTxErr blk))]
-implAddTxs mpEnv@MempoolEnv{mpEnvStateVar, mpEnvLedgerCfg, mpEnvTracer} txs = do
-    (removed, accepted, rejected, mempoolSize) <- atomically $ do
+implAddTxs mpEnv accum txs = do
+    (vr, removed, rejected, unvalidated, mempoolSize) <- atomically $ do
+      IS{isTip = initialISTip} <- readTVar mpEnvStateVar
+
       -- First sync the state, which might remove some transactions
-      syncRes@ValidationResult { vrInvalid = removed } <- validateIS mpEnv
-      -- Then validate the new transactions
-      let ValidationResult
-            { vrBefore, vrValid, vrLastTicketNo
-            , vrNewValid = accepted
-            , vrInvalid  = rejected
-            } = validateNew syncRes
-      writeTVar mpEnvStateVar IS { isTxs          = vrValid
-                                 , isTip          = vrBefore
-                                 , isLastTicketNo = vrLastTicketNo
-                                 }
-      mempoolSize <- getMempoolSize mpEnv
-      return (removed, accepted, rejected, mempoolSize)
+      syncRes@ValidationResult
+        { vrBefore
+        , vrValid
+        , vrInvalid = removed
+        , vrLastTicketNo
+        } <- validateIS mpEnv
+
+      -- Determine whether the tip was updated after a call to 'validateIS'
+      --
+      -- If the tip was updated, instead of immediately going on to call
+      -- 'validateNew' which can potentially 'retry', we should commit this
+      -- STM transaction to ensure that we don't lose any of the changes
+      -- brought about by 'validateIS' thus far.
+      --
+      -- On the other hand, if the tip wasn't updated, we don't have to worry
+      -- about losing any changes in the event that we have to 'retry' this
+      -- STM transaction. So we should continue by validating the provided new
+      -- transactions.
+      if initialISTip /= vrBefore
```

Comment:
This is just a realisation, not a comment on this code:
We have a background thread watching the chain and revalidating the Mempool as soon as the chain changes. So I think the chance is small that we actually end up in this case. Still, we need to handle it correctly here (which is what you're doing).